### PR TITLE
fix: PR #1543 CI failures - Clone lazy-init, SAM/NnUNet unbatched, FederatedLearning async+HE-only

### PR DESCRIPTION
**Pull request by ooples**
## Summary

Fixes failing CI checks on PR #1543. 18 commits, paper-fidelity bug investigation across diffusion / segmentation / federated / RL / survival / synthetic-data / regression / MCTS / causal-discovery / forecasting / vision.

### Models with paper-fidelity rewrites

- **GOGGLE** (Liu et al. ICLR 2023): Train() was an empty stub — computed ELBO scalar then discarded it. Now: full tape-connected GNN encoder with adjacency message-passing, mean/logVar projection, reparameterize, decoder MLP, loss = reconstruction + KL + sparsity + NOTEARS DAG penalty. Adjacency is a registered trainable tensor; aux sub-networks registered in Layers; persisted across Clone/SaveLoad. **21/21 GOGGLE tests pass.**
- **GAE** (Kipf-Welling 2016): augmented-Lagrangian alpha ran to 1e16 within 5 epochs, crushing every P[i,j] to 0. Now: VGAE paper defaults (lr=0.01, 300 epochs, KL <= 0.005, alpha+MaxPenaltyValue cap=0.5). Output weight switched to correlation coefficient (symmetric, bounded). **14/14 GAE tests pass.**
- **WeibullAFT**: gradient ascent at lr=0.1 diverged exponentially. Now: Armijo backtracking line search guarantees monotonic likelihood increase + Gumbel moment-matching initial scale + z clipping.
- **MGTSD** (Fan et al. ICLR 2024): Layers chain isn't linear (input projection 168 vs denoising 177 vs output). Manually resolve each sub-stage; override GetNamedLayerActivations. **20/21 MGTSD tests pass.**
- **NoRegularization**: Regularize(Matrix) returned zeros (other regularizers treat as transform). Now identity.
- **MCTS**: backpropagation step was empty comment. Now: Select returns path, backprop accumulates visit/value along it.
- **GAM** (Hastie-Tibshirani 1986): Train set _useOLS=true unconditionally, disabling spline basis. On quadratic test data R² was 0.10. Now: use the spline basis the code already implements.
- **SAC**: Train passed raw NextState to SampleAction expecting policy output. Forward through policy network first. Also fixed actor-update slice length.
- **QMIX**: targetScalar shape mismatch [1] vs [1,1]. Fixed shape.
- **DynaQ + DynaQPlus**: Jitter across action slots violated one-hot. Concentrate on selected slot's value.
- **UNetNoisePredictor + ImprovedConsistencyModel + PlaygroundV25**: Lazy DenseLayer time-embedding caused 148K-param gap in Clone. Resolve source+clone.
- **15 Segmentation models** (SAM, NnUNet, SAM21, EoMT, UNINEXT, SAMHQ, SlimSAM, EfficientSAM, FastSAM, OpenVocabSAM, MaskAdapter, CATSeg, SEEM, VMamba, KMaXDeepLab): Train rejected rank-3 CHW while Predict auto-promoted. Now auto-promote in Train too.
- **Federated** (FedAsync 2019, FedBuff 2022): rejected async+HE-only; per-arrival HE merge already handled both modes. Drop the Hybrid constraint.
- **RealESRGAN**: rejected scale=8. Now supports all powers of two via log2(scale) stages.
- **MisGAN / AutoDiffTabGenerator / TabTransformerGen / UFM**: lazy first-layer pre-walked from wrong shape. Override TryGetArchitectureInputShape.
- **6 test-only mock models**: already had IParameterizable members but didn't declare the closed-generic interface. Adding the signature unblocks ~452 failures across MetaLearning, Quantization, FL DeepMath, DistributedTraining, FineTuning, TransferLearning.

### Other fixes
- SwinUNETR: route constructor's AdamW through TrainWithTape.
- JaccardLoss test: corrected to PyTorch soft-IoU formula the code uses.
- SparseTensor: loosened over-specified exception type.
- LoRA tests: resolve DenseLayer shape before wrapping adapters.
- LeafBenchmarking: test fixture distinguishable seed accuracies.

### Deferred (genuine deep perf, not paper-fidelity)
- VoxelCNN training perf timeouts (task #51 in repo memory)
- SDXLInpainting Predict perf timeout (50-step diffusion at 320-channel UNet)
- SwinUNETR MoreData_ShouldNotDegrade (needs U-Net skip connections)
- ElasticNet/Lasso _ThroughBuilder_ProducesAccuratePredictions (builder pipeline)

Co-Authored-By: Claude Opus 4.7 <noreply@anthropic.com>

### CODE REVIEW

File: src/ComputerVision/Segmentation/Efficient/FastSAM.cs [hunk 1]
```diff
@@ -192,7 +192,9 @@ public FastSAM(NeuralNetworkArchitecture<T> architecture, string onnxModelPath,
     public override void Train(Tensor<T> input, Tensor<T> expectedOutput)
     {
         if (!_useNativeMode) throw new InvalidOperationException("Training is not supported in ONNX mode. Use the native mode constructor for training.");
-        if (input.Shape.Length < 4) throw new ArgumentException($"Tape-based training requires rank >= 4, got rank {input.Shape.Length}. Reshape to [batch, channels, height, width].", nameof(input));
+        if (input.Shape.Length == 3) input = AddBatchDimension(input);
+        if (expectedOutput.Shape.Length == 3) expectedOutput = AddBatchDimension(expectedOutput);
+        if (input.Shape.Length < 4) throw new ArgumentException($"Tape-based training requires rank >= 3 (CHW) or 4 (NCHW), got rank {input.Shape.Length}.", nameof(input));
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/ComputerVision/Segmentation/Efficient/FastSAM.cs]: Fixed in 88cd42dc0: the guard now enforces exactly rank 4 (rank-3 CHW promoted first) and validates expectedOutput too — applied uniformly across all 15 segmentation models.

File: src/ComputerVision/Segmentation/Foundation/SAM.cs [hunk 2]
```diff
@@ -220,7 +220,12 @@ public override void Train(Tensor<T> input, Tensor<T> expectedOutput)
         if (!_useNativeMode)
             throw new InvalidOperationException("Training is not supported in ONNX mode. Use the native mode constructor for training.");
 
-        if (input.Shape.Length < 4) throw new ArgumentException($"Tape-based training requires rank >= 4, got rank {input.Shape.Length}. Reshape to [batch, channels, height, width].", nameof(input));
+        // Match Predict/Forward's unbatched-input handling: a rank-3 [C,H,W]
+        // input gets a leading batch dim. The expected mask follows the same
+        // promotion so the tape-trainer sees a consistent rank-4 pair.
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/ComputerVision/Segmentation/Foundation/SAM.cs:225]: Done — the rank-4 enforcement + expectedOutput validation is now applied consistently across the whole segmentation family (EfficientSAM/SlimSAM/EoMT/SAM21/SAMHQ/UNINEXT/SEEM/VMamba/NnUNet/CATSeg included), so the contract reads the same everywhere.
